Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/AsyncServer.swift

```diff
@@ -0,0 +1,106 @@
+  private let server: EventLoopFuture<Server>
+  private let threadCount: Int
+
+  private var statsPeriodStart: DispatchTime
+  private var cpuStatsPeriodStart: CPUTime
+
+  private let logger = Logger(label: "AsyncQPSServer")
+
+  /// Initialisation.
+  /// - parameters:
+  ///     - config: Description of the type of server required.
+  ///     - whenBound: Called when the server has successful bound to a port.
+  init(config: Grpc_Testing_ServerConfig, whenBound: @escaping (ServerInfo) -> Void) {
+    // Setup threads as requested.
+    let threadCount = config.asyncServerThreads > 0 ? Int(config.asyncServerThreads) : System
+      .coreCount
+    self.threadCount = threadCount
+    self.logger.info("Sizing AsyncQPSServer", metadata: ["threads": "\(threadCount)"])
+    self.eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: threadCount)
+
+    // Start stats gathering.
+    self.statsPeriodStart = grpcTimeNow()
+    self.cpuStatsPeriodStart = getResourceUsage()
+
+    let workerService = AsyncQPSServerImpl()
+
+    // Start the server.
+    // TODO: Support TLS is requested.
+    self.server = Server.insecure(group: self.eventLoopGroup)
+      .withServiceProviders([workerService])
+      .withLogger(self.logger)
+      .bind(host: "localhost", port: Int(config.port))
+
+    self.server.whenSuccess { server in
+      let port = server.channel.localAddress?.port ?? 0
+      whenBound(ServerInfo(threadCount: threadCount, port: port))
+    }
+  }
+
+  /// Send the status of the current test
+  /// - parameters:
+  ///     - reset: Indicates if the stats collection should be reset after publication or not.
+  ///     - context: Context to describe where to send the status to.
+  func sendStatus(reset: Bool, context: StreamingResponseCallContext<Grpc_Testing_ServerStatus>) {
+    let currentTime = grpcTimeNow()
+    let currentResourceUsage = getResourceUsage()
+    var result = Grpc_Testing_ServerStatus()
+    result.stats.timeElapsed = (currentTime - self.statsPeriodStart).asSeconds()
+    result.stats.timeSystem = currentResourceUsage.systemTime - self.cpuStatsPeriodStart
+      .systemTime
+    result.stats.timeUser = currentResourceUsage.userTime - self.cpuStatsPeriodStart.userTime
+    result.stats.totalCpuTime = 0
+    result.stats.idleCpuTime = 0
+    result.stats.cqPollCount = 0
+    self.logger.info("Sending response")
+    _ = context.sendResponse(result)
+    if reset {
+      self.statsPeriodStart = currentTime
+      self.cpuStatsPeriodStart = currentResourceUsage
+    }
+  }
+
+  /// Shutdown the service.
+  /// - parameters:
+  ///     - callbackLoop: Which eventloop should be called back on completion.
+  /// - returns: A future on the `callbackLoop` which will succeed on completion of shutdown.
+  func shutdown(callbackLoop: EventLoop) -> EventLoopFuture<Void> {
+    return self.server.flatMap { server in
+      server.close()
+    }.hop(to: callbackLoop).flatMap { _ in
```

Comment:
We should probably stick a `recover` in here too in case the server errors on close.